### ISSUE TITLE
Enhancement: [no-unnecessary-condition] don't error on "exhaustiveness" checks

### ISSUE THREAD

**Issue by Retsam** [typescript-eslint/typescript-eslint#5836]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unnecessary-condition/

### Description

A common pattern that raises errors with `no-unnecessary-condition` is what I'm calling "exhaustiveness checking": (open to other names, too)

```ts
type Shape =
  | { kind: "square", side: number }
  | { kind: "circle", radius: number };

function assertNever(val: never, msg: string): never {
   throw new Error(msg);
}

function calculateArea(shape: Shape) {
   if(shape.kind === "square") { return shape.side * shape.side; }
   if(shape.kind === "circle") { return shape.radius * shape.radius * 3; }
   throw new Error("Unexpected shape");
   // or
   assertNever(shape, "Unexpected shape");
}
```
In general this is usually defensive against other variants being added - in particular the `assertNever` version, which raises an obvious compiler error when new variants are added but not handled - but is considered an error by `no-unnecesary-conditional`.  

I think it's worth carving out a special-case exception for if-checks where the alternative is to throw an error. 

### Fail

```typescript
type Shape =
  | { kind: "square", side: number }
  | { kind: "circle", radius: number };

function calculateArea(shape: Shape) {
   if(shape.kind === "square") { return shape.side * shape.side; }
   if(shape.kind === "circle") { return shape.radius * shape.radius * 3; } 

   doSomething();
}
```

### Pass

```typescript
type Shape =
  | { kind: "square", side: number }
  | { kind: "circle", radius: number };

function assertNever(val: never, msg: string): never {
   throw new Error(msg);
}

function calculateArea(shape: Shape) {
   if(shape.kind === "square") { return shape.side * shape.side; }
   if(shape.kind === "circle") { return shape.radius * shape.radius * 3; }
   throw new Error("Unexpected shape");
}

function calculateArea2(shape: Shape) {
   if(shape.kind === "square") { return shape.side * shape.side; }
   if(shape.kind === "circle") { return shape.radius * shape.radius * 3; }
   assertNever(shape, "Unexpected shape");
}
```

### Additional Info

I've been working on code for this, the current plan was to keep the exception fairly narrow and just check for either a throw or a `never` returning function as the "alternate" to the redundant if check.

**Comment by bradzacher** [comment]
Duplicate of #4682

In that issue outlined why I think this isn't a good idea, IMO, but TL;DR - the logic for the control-flow analysis required to support this is going to be pretty complex, and even then there's a wide potential for false-**_negatives_**.

---

Personally I suggest using a switch for these types of exhaustive checks.
Why? Because you can pair it with [`switch-exhaustiveness-check`](https://typescript-eslint.io/rules/switch-exhaustiveness-check) so that you can have errors without explicitly relying on the "assert never" pattern.
[Your example code, but with a switch and the rule](https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=C4TwDgpgBAygFgQ0lAvAWAFBSgHygbygGsBLAOwBMAuKAIgGcBHAVwQCcJaAaKekiiDTLMAtgCMIbKAF9M2PIVKUatAMYk2qgDacebBBRLN6Q0RKmysuAsXLU6wNiQRkA5ju5Q4EEq7jBTcUkeMQR6QShhIIsAbkxMADNmMlVgEgB7MihVBC1VZi0EYAgAQQ4EAHF09IoACnpESBp4JAgASgI5XgB3EmBVOCh6xogAOiUKDvwu7BzwuiZWDloaDmBmNiyG1tG+ASgAKl4R3f4IOKtZsOg1DW1OVYh1zeOd-UNjQ9fIUfejei+AGYLtgrvNaI5nG4PI9nlsTqF5kdtj9vL5-F8AAyjACsIJkmEs8QwSRSaUy2Vy+UKxTKEAQACEDMNWs0RlMuvRev1BiyfhMOZdKeDFuwHlA1ht4Ts9tBkSdZfiwTd1JoYRKnlLvmM-p95W8DP8gfjLESMIlkqkMlkwuE2MAAHIQABuklqztyQhdwSgInorho9EhbjaXtdUmmVmAcDY6W6kQg8YAomxY2xan7XG0LpZSVaKTk8gUiqVykyKAAmPkRFqQQXYEgJIYosYTVAoFALFhi2hTDVw7Wnfb6n6KglWRvNk5tjud25qzh9yUvFu-Q16we6gFHYHj7DR2PxsiJqAptO1WgAVWPAA9IKkIBRtb2cxaydbKUWaaX6YDq2zWnrKBJ2rcY7HbOdRWWJdNRXBUzi+Vcx0sBsm1AmcOzoVV7l7Gxl2lH4t0Qk4iJ3GI9ygW1JEdb10xbHgr1ve9iifFsX0JIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6KImjQA9tEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

With the "if chain plus throw error" pattern - it's really bad to do that in TS because it's a silent error - TS will not catch it, nor can any lint rules. At least with the "if chain plus assert never" pattern TS will catch it, but the error (IMO) is a little cryptic - certainly less clear than the error from the switch lint rule.

**Comment by Retsam** [comment]
Ah, forgot to check the closed issues filter, too.   

> TL;DR - the logic for the control-flow analysis required to support this is going to be pretty complex, and even then there's a wide potential for false-negatives.

I don't think it's actually going to be that much logic.  I have it already implemented for the basic case and it's ~60 heavily commented lines.  (And I think I might be reinventing some wheels that can be replaced with built-in eslint code path utilities)  

And I'm planning to keep it fairly targeted to hopefully avoid false negatives: it's a pretty stock pattern, I don't think the logic needs to be particularly flexible.  

---

And I'm not a big fan of the switch version of this.  For one, I just find switches more syntactically awkward than if/else blocks (extra indentation, breaks, etc); but it's also just less flexible - sometimes logic is shared between multiple cases, e.g.

```ts
type RawShape = { kind: "circle",  radius: string } | { kind: "square", width: string } | { kind: "rectangle", width: string, height: string }

function calculateArea(shape: RawShape) {
    if(shape.kind === "circle") { 
        const radius = parseInt(shape.radius);
        return radius * radius * 3;
    }
    // shared logic for other cases
    const width = parseInt(shape.width);
    if(shape.kind === "square") {
        return width * width;
    }
    if(shape.kind === "rectangle") {
        const height = parseInt(shape.height);
        return width * height;
    }
    assertNever(shape, "oops");
}
```

This is obviously, super contrived, but I've definitely done this sort of logic with React/Redux reducer functions.

And I don't think `assertNever` is  a particularly cryptic pattern: [it's came directly out of the handbook](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-exhaustiveness-checking).  (This is the old handbook version, I'd assume it's still somewhere in the new handbook, but I'm less familiar with it)

Comment overrides aren't the end of the world, but they do create the sort of friction that encourages people to turn off rules.  (And not everybody feels that way about comments - I was just talking the other day with someone who said they keep them out of their code base, essentially as a rule)  

If this rule can handle this case without a ton of work, I think it should.  The false negative thing is a fair point, but I think it's avoidable.
